tactics/rewrite.ml

```diff
@@ -1344,49 +1346,70 @@ module Strategies =
       let rec aux input = (f { strategy = fun input -> check_interrupt aux input }).strategy input in
       { strategy = aux }
 
+    let fix_tac (f : 'a pure_strategy -> 'a pure_strategy Proofview.tactic) : 'a pure_strategy Proofview.tactic =
+      let forward_def = ref (fun _ -> assert false) in
+      f {strategy = fun input -> check_interrupt !forward_def input} >>= fun f ->
+      forward_def := f.strategy;
+      Proofview.tclUNIT f
+
+    let all_subterms (s : strategy) : strategy = subterm true default_flags s
+
+    let one_subterm (s : strategy) : strategy = subterm false default_flags s
+
     let any (s : 'a pure_strategy) : 'a pure_strategy =
       fix (fun any -> try_ (seq s any))
 
     let repeat (s : 'a pure_strategy) : 'a pure_strategy =
       seq s (any s)
 
-    let bu (s : 'a pure_strategy) : 'a pure_strategy =
+    let bottomup (s : strategy) : strategy =
       fix (fun s' -> seq (choice (progress (all_subterms s')) s) (try_ s'))
 
-    let td (s : 'a pure_strategy) : 'a pure_strategy =
+    let topdown (s : strategy) : strategy =
       fix (fun s' -> seq (choice s (progress (all_subterms s'))) (try_ s'))
 
-    let innermost (s : 'a pure_strategy) : 'a pure_strategy =
+    let innermost (s : strategy) : strategy =
       fix (fun ins -> choice (one_subterm ins) s)
 
-    let outermost (s : 'a pure_strategy) : 'a pure_strategy =
+    let outermost (s : 'a pure_strategy) : strategy =
       fix (fun out -> choice s (one_subterm out))
 
-    let lemmas cs : 'a pure_strategy =
-      List.fold_left (fun tac (l,l2r,by) ->
-        choice tac (apply_lemma l2r rewrite_unif_flags l by AllOccurrences))
+    let one_lemma c l2r by occs : strategy =
+      let strategy ({env} as input) =
+        let c sigma = with_no_bindings c env sigma in
+        let flags = general_rewrite_unif_flags () in
+        (apply_lemma l2r flags c by occs).strategy input
+      in {strategy}
+
+    let lemmas cs : strategy =
+      List.fold_left (fun tac (l,l2r,by) -> choice tac (one_lemma l l2r by AllOccurrences))
         fail cs
```

Comment:
For some reason, `terms` used glob constr while `term` used delayed econstr. I couldn't find a reason for this inconsistency and thus made both work on econstr (but glob should be fine as well).